mindmeld/text_preparation/tokenizers.py

```diff
@@ -0,0 +1,258 @@
+            actions_by_char.append(action)
+        return actions_by_char
+
+    @staticmethod
+    def get_token_num_by_char(actions_by_char):
+        """
+        Get the token number that each character belongs to.
+        Args:
+            actions_by_char (List[str]): List of actions corresponding to each character.
+                For example: ["separate", "separate", "skip", "no_action"]
+        Returns:
+            token_num_by_char (List[str]): Token number that each character belongs to.
+                Spaces are represented as -1. For example: [1,2,2,3,-1,4,-1,5,5,5]
+        """
+        token_num_by_char = []
+        token_num = 0
+        for action in actions_by_char:
+            if action == "skip":
+                token_num_by_char.append(-1)
+                continue
+            if action == "separate":
+                token_num += 1
+            token_num_by_char.append(token_num)
+        return token_num_by_char
+
+    @staticmethod
+    def create_tokens(text, token_num_by_char):
+        """
+        Generate token dictionaries from the original text and the token numbers by character.
+        Args:
+            text (str): The text to normalize
+            token_num_by_char (List[str]): Token number that each character belongs to.
+                Spaces are represented as -1. For example: [1,2,2,3,-1,4,-1,5,5,5]
+        Returns:
+            tokens (List[Dict]): List of tokenized tokens which a represented as dictionaries.
+                Keys include "start" (token starting index), "end" (token ending index), and
+                "text" (token text). For example: [{"start": 0, "text":"hello", "end":4}]
+        """
+        tokens = []
+        token_text = ""
+        for index, token_num in enumerate(token_num_by_char):
+            if token_num == -1:
+                continue
+            if not token_text:
+                start = index
+            token_text += text[index]
+            is_last_char = index == len(token_num_by_char) - 1
+            # Close off entity if char is the last or if next char is a different token number
+            if is_last_char or (
+                not is_last_char and token_num != token_num_by_char[index + 1]
+            ):
+                tokens.append({"start": start, "text": token_text, "end": index})
+                token_text = ""
+        return tokens
+
+
+class WhiteSpaceTokenizer(Tokenizer):
+    """A Tokenizer that splits text at spaces."""
+
+    def __init__(self):
+        """Initializes the WhiteSpaceTokenizer."""
+        pass
+
+    def tokenize(self, text):
+        """
+        Identify tokens in text and create normalized tokens that contain the text and start index.
+        Args:
+            text (str): The text to normalize
+        Returns:
+            tokens (List[str]): List of tokenized tokens.
+        """
+        tokens = []
+        token = {}
+        token_text = ""
+        for i, char in enumerate(text):
```

Comment:
You could do this `for i, char in enumerate(text + ' '):` and then 194-197 will not be required

Reply:
Good point! Made this change and added a comment explaining the added space.